server-timing/class-perflab-server-timing.php

```diff
@@ -0,0 +1,255 @@
+
+	/**
+	 * Map of registered metric slugs and their registered data.
+	 *
+	 * @since n.e.x.t
+	 * @var array
+	 */
+	private $registered_metrics_data = array();
+
+	/**
+	 * Registers a metric to calculate for the Server-Timing header.
+	 *
+	 * This method must be called before the {@see 'perflab_server_timing_send_header'} hook.
+	 *
+	 * @since n.e.x.t
+	 *
+	 * @param string $metric_slug The metric slug.
+	 * @param array  $args        {
+	 *     Arguments for the metric.
+	 *
+	 *     @type callable $measure_callback The callback that initiates calculating the metric value. It will receive
+	 *                                      the Perflab_Server_Timing_Metric instance as a parameter, in order to set
+	 *                                      the value when it has been calculated. Metric values must be provided in
+	 *                                      milliseconds.
+	 *     @type string   $access_cap       Capability required to view the metric. If this is a public metric, this
+	 *                                      needs to be set to "exist".
+	 * }
+	 */
+	public function register_metric( $metric_slug, array $args ) {
+		if ( isset( $this->registered_metrics[ $metric_slug ] ) ) {
+			_doing_it_wrong(
+				__METHOD__,
+				/* translators: %s: metric slug */
+				sprintf( __( 'A metric with the slug %s is already registered.', 'performance-lab' ), $metric_slug ),
+				''
+			);
+			return;
+		}
+
+		if ( did_action( 'perflab_server_timing_send_header' ) && ! doing_action( 'perflab_server_timing_send_header' ) ) {
+			_doing_it_wrong(
+				__METHOD__,
+				/* translators: %s: WordPress action name */
+				sprintf( __( 'The method must be called before or during the %s action.', 'performance-lab' ), 'perflab_server_timing_send_header' ),
+				''
+			);
+			return;
+		}
+
+		$args = wp_parse_args(
+			$args,
+			array(
+				'measure_callback' => null,
+				'access_cap'       => null,
+			)
+		);
+		if ( ! $args['measure_callback'] || ! is_callable( $args['measure_callback'] ) ) {
+			_doing_it_wrong(
+				__METHOD__,
+				/* translators: %s: PHP parameter name */
+				sprintf( __( 'The %s argument is required and must be a callable.', 'performance-lab' ), '$args["measure_callback"]' ),
+				''
+			);
+			return;
+		}
+		if ( ! $args['access_cap'] || ! is_string( $args['access_cap'] ) ) {
+			_doing_it_wrong(
+				__METHOD__,
+				/* translators: %s: PHP parameter name */
+				sprintf( __( 'The %s argument is required and must be a string.', 'performance-lab' ), '$args["access_cap"]' ),
+				''
+			);
+			return;
+		}
+
+		$this->registered_metrics[ $metric_slug ]      = new Perflab_Server_Timing_Metric( $metric_slug );
+		$this->registered_metrics_data[ $metric_slug ] = $args;
+
+		// If the current user has already been determined and they lack the necessary access,
+		// do not even attempt to calculate the metric.
+		if ( did_action( 'set_current_user' ) && ! current_user_can( $args['access_cap'] ) ) {
```

Comment:
I assume this is to prevent non user invocation - out of curiosity, when is `did_action( 'set_current_user' )` not true? during something like a cron job, or even on unauthenticated requests? why not a more explicit call like `wp_get_current_user`?

Reply:
@adamsilverstein The current user is initialized right before `init`. This means for example when calling this on `plugins_loaded` it won't be initialized.

It is certainly possible to initialize the current user earlier than that (by calling any function that relies on the current user), but that's not a great pattern and can lead to other unexpected behavior.

Related, see my response in https://github.com/WordPress/performance/pull/553#discussion_r1048634541 in regards to this check not being critical anyway - it's just a nice to have.

Reply:
👍🏼